### PR TITLE
feat(tooling): repair_db — rebuild MANIFEST from on-disk SSTs

### LINKED ISSUE #303: feat(tooling): repair_db — rebuild MANIFEST from SST files (disaster recovery)
## Problem

Once we have a MANIFEST, it becomes a single point of failure for the DB as a whole: corrupt MANIFEST = can't open the DB at all, even if every SST on disk is intact. Standard recovery in RocksDB is `RepairDB()` — scan the directory for SST files, read each one's meta block, reconstruct a fresh MANIFEST from what's actually present on disk. Loses recent VersionEdits (compactions in flight, deletions not yet logged) but the data is recoverable.

## Solution

`Tree::repair(path)` function and corresponding CLI subcommand:

1. Scan directory for `*.sst` files
2. Open each SST, read its meta block (table_id, key_range, level hints if present)
3. Group by level (best-effort heuristic — newest files are typically newer levels) OR put all at L0 + force a re-compaction
4. Emit a fresh MANIFEST with one VersionEdit per discovered SST
5. Report: N files recovered, M files unreadable (skipped), warnings about ambiguous level assignments

## Acceptance

- [ ] `Tree::repair(path) -> Result<RepairReport>` API
- [ ] CLI subcommand `sst-dump repair <db-dir>` (extending #301)
- [ ] Integration test: delete MANIFEST, run repair, verify all data still readable via point reads on known keys
- [ ] Repair report documents what was lost (recent VersionEdits, level assignments)

## Estimate

3-4d (when MANIFEST work lands)

**Blocked by:** MANIFEST design / implementation. Related: matches RocksDB `RepairDB()`.

---

## Decision (Q19): Level assignment — all-to-L0 + force recompaction (RocksDB pattern)

**All recovered SSTs placed at L0 with sequence-number ordering. Tree triggers background compaction to restructure into proper LSM levels.**

### Implementation

```rust
pub fn repair(path: &Path) -> Result<RepairReport> {
    let sst_files = scan_sst_files(path)?;
    let mut version_edits = Vec::new();
    let mut unreadable = Vec::new();

    for sst_path in sst_files {
        match open_sst_get_metadata(&sst_path) {
            Ok(meta) => {
                version_edits.push(VersionEdit::AddTable {
                    id: meta.id,
                    smallest_key: meta.smallest,
                    largest_key: meta.largest,
                    sequence_number: meta.seqno,
                    level: 0,  // ALL-to-L0
                });
            }
            Err(e) => unreadable.push((sst_path, e)),
        }
    }

    // Sort by sequence number — newer SSTs near L0 head
    version_edits.sort_by_key(|e| std::cmp::Reverse(e.sequence_number()));

    write_manifest(path, &version_edits)?;
    schedule_post_repair_compaction();

    Ok(RepairReport {
        recovered: version_edits.len(),
        unreadable: unreadable.len(),
        unreadable_files: unreadable.into_iter().map(|(p, e)| (p, e.to_string())).collect(),
        method: "all-to-L0 with sequence-number ordering",
        warnings: vec![
            "All recovered files placed at L0; background compaction will redistribute",
            "Expect elevated I/O for hours to days on large DBs as compaction catches up",
        ],
    })
}
```

### RepairReport

```rust
pub struct RepairReport {
    pub recovered: usize,
    pub unreadable: usize,
    pub unreadable_files: Vec<(PathBuf, String)>,
    pub method: &'static str,
    pub warnings: Vec<&'static str>,
}
```

### Why A over alternatives

**A over C (key-range overlap analysis):** C optimizes placement but recovery is a rare last-resort path; complex level-packing logic earns small payback for high test-surface and bug risk. Simple > clever for disaster code.

**A over B (mtime heuristic):** mtime is unreliable across recovery scenarios (cp without `-p`, restore tools, backup pipelines). Recovery should trust only SST content, not filesystem metadata.

**A over D (hybrid):** falls back to A anyway — no value in conditional complexity.

### Documentation caller-facing message

> `Tree::repair()` places all recovered SSTs at L0 and triggers background compaction to rebuild the LSM level structure. This preserves all data but produces elevated I/O for hours to days on large databases as compaction redistributes files into proper levels. The tree is fully operational throughout — reads work correctly during the catch-up period, just at reduced performance until L0 size returns to normal.
>
> For databases > 100 GB, expect 1-3 days of background recompaction following repair. Schedule operator monitoring accordingly.

### Comparison

| System | Repair API | Level assignment |
|--------|------------|------------------|
| RocksDB `RepairDB()` | yes | All-to-L0 + recompact (this pattern) |
| Pebble | none | n/a — distributed layer handles it |
| LMDB / MDBX | n/a (single file) | n/a |
| **Us (Q19)** | **yes** | **All-to-L0 + recompact (RocksDB parity)** |

### Marketing positioning

"Last-resort manifest recovery with full data preservation. All-to-L0 placement + automatic background recompaction matches industry-standard RocksDB pattern. Tree fully operational immediately; optimal layout restored via compaction over hours to days proportional to data size."

vs RocksDB: parity (same algorithm).
vs Pebble: superior — Pebble ships no repair API.

### Benchmark Symmetry compliance (#353)

N/A — repair_db is an operational disaster-recovery tool, not a benchmark target.

### Estimate

~150 LOC implementation + ~100 LOC tests = ~250 LOC. ~2 days wall time.

Implementation can land post-V5 (depends on #352/#297 manifest format being stable). Not in V5 batch.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Two related pieces:

1. **`repair_db`** — last-resort `MANIFEST` reconstruction from the SST files on disk (`Config::repair()` + `sst-dump <db-dir> repair`).
2. **structured-zstd 0.0.29 bump + encoder-side dictionary / compressor-reuse adoption** — picked up while the repair path exercises the compression stack; roughly halves our zstd:22 encode time.

## repair_db

- `Config::repair() -> Result<RepairReport>`: scans the table folder(s), recomputes each SST's whole-file XXH3-128 checksum, validates readability via `Table::recover`, and writes a fresh manifest placing every recovered table at L0 (RocksDB `RepairDB()` pattern; background compaction restructures on next open).
- Unreadable files are reported and skipped, not fatal. KV-separated (blob) trees return `FeatureUnsupported` (tracked in #408).
- `sst-dump <db-dir> repair` CLI wrapper.

## Compression (structured-zstd 0.0.29)

- `build(deps)`: 0.0.28 → 0.0.29 (per-strategy block-split, decode-bomb output bound, donor-correct block-split).
- **A**: attach the encode dictionary via `EncoderDictionary` (`set_dictionary_from_bytes` / `set_encoder_dictionary`) — parses for the encoder side only, skipping the decode lookup tables the compressor never reads.
- **B**: reuse a thread-local `FrameCompressor` across block frames via `compress_independent_frame` — avoids reconstructing matcher tables per block; the independent-frame path re-derives the source-size hint so the L22 small-source parameter set is still selected.

### Measured (local compare-rocksdb, zstd:22, median)

| bench | ours | rocksdb | A+B delta |
|-------|------|---------|-----------|
| write_throughput / 10k | 3.38 s | 4.11 s | ~2x faster |
| overwrite / 10k | 2.92 s | 4.11 s | ~2x faster |
| major_compact / 10k | 0.87 s | 1.06 s | ~2x faster |

Our zstd:22 encode now beats RocksDB on these. (zstd:1 also improved ~30% but still trails RocksDB; the subcompaction gap is split out as #410.)

## Testing

- repair: manifest-loss round-trip (all keys readable after repair incl. overwrites), unreadable-file skip, non-table-id / dangling-symlink / .DS_Store handling, zero-SST, blob-tree rejection; CLI smoke test.
- compression: dictionary round-trip / wrong-dict rejection / per-level / encryption / blob / major-compaction (39 tests) unchanged; full workspace suite green; clippy `--all-features` + fmt clean; doc tests pass.

## Related

- #408 — repair for KV-separated (blob) trees (follow-up)
- #410 — subcompaction far slower than RocksDB (perf investigation, split out)

Closes #303

## Summary by CodeRabbit

* **New Features**
  * Added a disaster-recovery repair mechanism to rebuild missing or corrupt manifests from on-disk SST files.
  * Added a CLI "repair" command to run reconstruction and print a repair report.

* **Behavior & Reliability**
  * Recovery skips or quarantines unreadable or non-table files, preserves readable data, and reports recovered/unreadable counts, per-file reasons, and warnings.
  * Rejects unsupported KV-separated/tree formats.

* **Tests**
  * Added unit and end-to-end tests covering repair scenarios and edge cases.

* **Improvements**
  * Improved compression performance by reusing compressors and refining dictionary handling.

### CODE REVIEW

**Comment by polaz** [comment]
@coderabbitai re-review

File: src/repair.rs [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+fn highest_existing_version_id(fs: &dyn crate::fs::Fs, folder: &std::path::Path) -> Option<u64> {
+    let entries = fs.read_dir(folder).ok()?;
+    entries
+        .into_iter()
+        .filter_map(|e| {
+            e.file_name
+                .strip_prefix('v')
+                .and_then(|rest| rest.parse::<u64>().ok())
+        })
+        .max()
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/repair.rs]: @coderabbitai Fixed in be8f84c8 — `highest_existing_version_id` now returns `Result<Option<u64>>` and propagates the `read_dir` error via `repair_tree`, so a transient scan failure no longer silently resets the version id to 0.

File: src/repair.rs [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+    crate::version::persist_version(
+        &config.path,
+        &version,
+        config.comparator.name(),
+        &*config.fs,
+        Arc::new(crate::runtime_config::RuntimeConfig::default()),
+        config.encryption.clone(),
+        config.sync_mode,
+    )?;
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/repair.rs:334]: @coderabbitai Fixed in be8f84c8 — repair now persists with `config.initial_runtime_config` instead of `RuntimeConfig::default()`, so the rebuilt manifest's framing (checksum algorithm, page ECC, footer mirror, manifest KV checksums) matches the recovered tree's configuration rather than synthesizing defaults.

File: tests/repair.rs [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+        let entry = entry.unwrap();
+        let name = entry.file_name();
+        let name = name.to_string_lossy();
+        let is_version = name
+            .strip_prefix('v')
+            .is_some_and(|rest| rest.parse::<u64>().is_ok());
+        if is_version || name == "current" {
+            std::fs::remove_file(entry.path()).unwrap();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/repair.rs]: @coderabbitai Fixed in be8f84c8 — `nuke_manifest` and `count_sst_files` now return `std::io::Result` and all filesystem I/O in the tests uses `?` instead of `unwrap()`.

File: tools/sst-dump/src/main.rs [hunk 4]
```diff
@@ -155,6 +155,21 @@ enum Command {
...
+    /// recovered table at L0 (background compaction restructures it on
+    /// the next open). Recent unlogged version edits (in-flight
+    /// compactions, recent deletions) are lost; all readable data is
+    /// preserved.
+    ///
+    /// Uses the default comparator and no encryption: a tree created
+    /// with a custom comparator or with encryption must be repaired via
+    /// the library `Config::repair` API with the matching config.
+    /// KV-separated (blob) trees are not supported by repair.
+    Repair,
```
[the remark this answers is not in the document]
Comment (polaz) [reply tools/sst-dump/src/main.rs:174]: @coderabbitai Fixed in be8f84c8 — the positional is renamed `file` → `path` and its help now states it's an SST file for the inspection subcommands and the DB directory for `repair`.

File: tools/sst-dump/tests/repair_smoke.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        let name = entry.file_name();
+        let name = name.to_string_lossy();
+        let is_version = name
+            .strip_prefix('v')
+            .is_some_and(|rest| rest.parse::<u64>().is_ok());
+        if is_version || name == "current" {
+            std::fs::remove_file(entry.path()).expect("remove manifest file");
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (polaz) [reply tools/sst-dump/tests/repair_smoke.rs]: @coderabbitai Fixed in be8f84c8 — the smoke test returns `Result<(), Box<dyn Error>>` and `nuke_manifest` returns `std::io::Result<()>`; all filesystem/process I/O uses `?` instead of `expect()`.

File: src/compression/zstd_backend.rs [hunk 6]
```diff
@@ -239,17 +239,38 @@ pub struct ZstdProvider;
...
+                *state = Some((
+                    level,
+                    FrameCompressor::new(CompressionLevel::from_level(level)),
+                ));
+            }
+            let Some((_, compressor)) = state.as_mut() else {
+                unreachable!("TLS_COMPRESSOR initialised above");
+            };
+            Ok(compressor.compress_independent_frame(data))
+        })
```
Comment (polaz) [inline src/compression/zstd_backend.rs:276]: @coderabbitai False positive — `FrameCompressor::compress_independent_frame(&mut self, input: &[u8]) -> Vec<u8>` **does** exist in structured-zstd 0.0.29 (`src/encoding/frame_compressor.rs:713`, alongside `compress_independent_frame_into` at :690). The web-query result was from a stale docs.rs version. Proof: the code compiles and the full workspace suite (1834 tests incl. 39 zstd dictionary round-trips + the compare-rocksdb zstd benches) is green on 0.0.29. Added a comment noting it's the 0.0.29 CCtx-style API (dd453bf3).

File: src/repair.rs [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+
+            // https://en.wikipedia.org/wiki/.DS_Store
+            if is_dir || file_name == ".DS_Store" || file_name.starts_with("._") {
+                continue;
+            }
+
+            let Ok(table_id) = file_name.parse::<TableId>() else {
+                unreadable_files.push((table_path, "file name is not a table id".to_owned()));
+                continue;
+            };
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/repair.rs:233]: @coderabbitai Fixed in 589a0cca — repair now moves non-table-id files out of `tables/` into a sibling `repair-quarantine/` dir (a plain rename; preserved, not deleted, since repair didn't create them) and reports the new location. The tree now reopens with no manual cleanup. Regression test `repair_reports_non_table_id_filename_as_unreadable` (01ab56cd) reopens after repair without removing the junk and asserts it was moved out of `tables/`.

File: tests/repair.rs [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+    .open()?;
+    for i in 0..100 {
+        assert_eq!(
+            tree.get(key(i), MAX_SEQNO)?.as_deref(),
+            Some(format!("v0-{i}").as_bytes()),
+        );
+    }
+
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/repair.rs:186]: @coderabbitai Added in 23f5801c — `repair_rejects_corrupted_sst_and_recovers_the_rest`: flushes two real SSTs, tampers the newest one's SFA trailer, and asserts `Table::recover` rejects it during repair (report.unreadable == 1, path matches) while the intact SST still recovers (report.recovered == total-1) and the tree reopens with the intact keys readable.
